### PR TITLE
Generator Spec

### LINKED ISSUE #63: Generator spec adjustments
- [ ] Generator manifest
- [ ] Binary requirements via generator manifest

### PR DESCRIPTION
**Pull request by timsuchanek**
[Rendered View](https://github.com/prisma/specs/tree/generators/generators)

Closes #63

### CODE REVIEW

File: generators/Readme.md [hunk 1]
```diff
@@ -0,0 +1,302 @@
+# Generators
+
+- Owner: @timsuchanek
+- Stakeholders: @schickling
+- State:
+  - Spec: In Progress 🚧
+  - Implementation: In Progress 🚧
+
+An important mechanism in Prisma 2 is the generation of code based on the Prisma schema.
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
An important mechanism in Prisma 2 is the generation of code based on the Prisma schema.  
```
to keep the line break in rendered view

File: generators/Readme.md [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+└─────────────────┘
+```
+
+## Information passed into a generator
+
+A generator gets passed in a JSON blob including the following information:
+
+1. Everything which is declared in the `schema.prisma` file:
+
+- Models
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
   - Models
```
to get the unordered list as a sub thing of the numbered list (also for all the other items of course)

File: generators/Readme.md [hunk 3]
```diff
@@ -0,0 +1,302 @@
+# Generators
+
+- Owner: @timsuchanek
+- Stakeholders: @schickling
+- State:
+  - Spec: In Progress 🚧
+  - Implementation: In Progress 🚧
+
+An important mechanism in Prisma 2 is the generation of code based on the Prisma schema.
+This spec describes, how the interface for such code generation in the individual languages like JavaScript and TypeScript looks like, and which protocol is used under the hood to achieve that interface.
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
This spec describes how the interface for such code generation in the individual languages like JavaScript and TypeScript looks like, and which protocol is used under the hood to achieve that interface.
```

File: generators/Readme.md [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+```
+
+</details>
+
+## Interface for a generator
+
+To add a generator to a Prisma project and run it, the following steps are needed:
+
+1. Create an executable file, which handles the generation
+2. Point to that file in the `schema.prisma`
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
2. Point to that file in your `schema.prisma`
```

File: generators/Readme.md [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+3. Run `prisma2 generate`
+
+### 1. Create an executable file
+
+In the following we describe how this can be done in JavaScript or TypeScript.
+In section [Interface between Generator SDK and each language](#interface-between-generator-sdk-and-each-language) you can read more about other languages.
+
+To create a generator in TypeScript, create a new package, which needs two files: A executable generator file, e.g. `generator.ts` and a generator manifest, called `generator-manifest.json`.
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
To create a generator in TypeScript, create a new package, which needs two files: A executable generator file, e.g. `generator.ts` and optionally a generator manifest, called `generator-manifest.json`.
```
Beneath this is defined as optional

File: generators/Readme.md [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+`Stderr` is being used to communicate back to the SDK:
+
+```ts
+{"jsonrpc": "2.0", "result": { error: "Could not find directory" }, "id": 1}
+```
+
+These are the available RPCs:
+
+1. `generate` With the input `GeneratorOptions`. If the generator process crashes or returns an object containing an `error` property, the generator SDK can handle that error and show it to the user.
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
1. `generate` with the input `GeneratorOptions`. If the generator process crashes or returns an object containing an `error` property, the generator SDK can handle that error and show it to the user.
```

**Comment by janpio** [comment]
Confused about "Generator SDK vs Generator vs Generation SDK".
Where does this SDK live? Is this part of Prisma? And the developer only creates a Generator?

**Comment by timsuchanek** [comment]
Good question! I added a terminology section to answer that question.

File: generators/Readme.md [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+
+## Information passed into a generator
+
+A generator gets passed in a JSON blob including the following information:
+
+1. Everything which is declared in the `schema.prisma` file:
+
+   - Models
+
+- Enums
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
   - Enums
```

File: generators/Readme.md [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+- Datasources
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
   - Datasources
```

File: generators/Readme.md [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+- Generators, including the configuration this generator points to
```
Comment (janpio) [inline generators/Readme.md]:
```suggestion
   - Generators, including the configuration this generator points to
```

**Comment by schickling** [comment]
# Feedback

- [ ] Generators are not just about generating code but generally generating artifacts (e.g. a uml-diagram.png). Please adjust the language accordingly

> A generator is a piece of software
- [ ] I’d rather say “an executable”

> The Generator SDK is part of the Prisma SDK
Not sure about that yet. We’ll get more clarity as we’ll get to the (eventual) redesign of the Prisma SDK

- [ ] The Terminology section should also contain an item I’d suggest calling a Generated artifact (short artifact) which is the thing that a Generator generates

> A Generator is created by a developer using Prisma.
- [ ] Could be more descriptive

> This part is still pretty vague and should have a standalone spec. @janpio  @timsuchanek is there one? (I guess that’s what we currently call `DMMF`?)
image

- [ ] Would be great if the spec could link to a workign example repo
image

**Comment by divyenduz** [comment]
# Feedback

- [ ] Should the `GeneratorConfig` options `platforms` and `pinnedPlatforms` be renamed to `binaryTargets` and `pinnedBinaryTarget` respectively for consistency? 
- [ ] The options from point 1 define which platform to target but we might need additional configuration for **what** binaries are required like `photon.js` requires just the query engine but `prisma-test-utils` requires both migration engine and query engine. Or I am not clear about the development workflow, how would generate know which binaries to download?

File: generators/Readme.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+- A `Generator` is created by a developer using Prisma.
+
+## Information passed into a generator
+
+A generator gets passed in a JSON blob including the following information:
+
+1. Everything which is declared in the `schema.prisma` file:
+
+   - Models
+
```
Comment (janpio) [inline generators/Readme.md]: get rid of this line ;)

**Comment by timsuchanek** [comment]
> This part is still pretty vague and should have a standalone spec. @janpio @timsuchanek is there one? (I guess that’s what we currently call DMMF?)

As you can see in the TypeScript definition, the DMMF is a part of it, but the "part" you refer to is called `GeneratorOptions`. How should we go more into detail than the typings? And yes, the pure DMMF part of it should be documented separately.

**Comment by timsuchanek** [comment]
Update:
- Added `binaryPaths` to the `GeneratorOptions`
- Added `requiresGenerators` and `requiresEngines` to the `generator-manifest.json`

File: generators/Readme.md [hunk 11]
```diff
@@ -6,15 +6,14 @@
   - Spec: In Progress 🚧
   - Implementation: In Progress 🚧
 
-An important mechanism in Prisma 2 is the generation of code based on the Prisma schema.  
-This spec describes how the interface for such code generation in the individual languages like JavaScript and TypeScript looks like, and which protocol is used under the hood to achieve that interface.
+An important mechanism in Prisma 2 is the generation of artifacts like based on the Prisma schema.
```
Comment (janpio) [inline generators/Readme.md]: like what?

File: generators/Readme.md [hunk 12]
```diff
@@ -64,23 +64,22 @@ A generator gets passed in a JSON blob including the following information:
...
 ```ts
 type GeneratorConfig = {
   output: string | null
   name: string
   provider: string
   config: Dictionary<string>
-  platforms: string[]
-  pinnedPlatform?: string | null
+  binaryTargets: string[]
+  pinnedBinaryTarget?: string | null
```
Comment (janpio) [inline generators/Readme.md]: This relates to some other spec (that was recently changed)? If so, linking there would be smart here.

**Pull request merged**